File: naive_bayes.py
import pandas as pd
from sklearn.model_selection import train_test_split
import os

# Read and preprocess data
def read_data(file_path):
    # Load data into a Pandas DataFrame
    df = pd.read_csv(file_path)
    
    # Select relevant columns
    negative_reviews = df[df['Negative_Review'] != 'No Negative'][['Negative_Review']]
    negative_reviews['sentiment'] = 0  # Label negative reviews

    positive_reviews = df[df['Positive_Review'] != 'No Positive'][['Positive_Review']]
    positive_reviews['sentiment'] = 1  # Label positive reviews

    # Combine positive and negative reviews
    reviews = pd.concat([negative_reviews.rename(columns={'Negative_Review': 'text'}),
                         positive_reviews.rename(columns={'Positive_Review': 'text'})])
    
    # Shuffle data and reset index
    reviews = reviews.sample(frac=1, random_state=42).reset_index(drop=True)
    return reviews['text'], reviews['sentiment']

# Tokenize text
def tokenize(text):
    return text.lower().split()

# Train Naive Bayes classifier
def train_naive_bayes(X_train, y_train):
    word_counts = {0: {}, 1: {}}
    class_counts = {0: 0, 1: 0}
    vocabulary = set()

    for text, label in zip(X_train, y_train):
        tokens = tokenize(text)
        class_counts[label] += len(tokens)
        for token in tokens:
            vocabulary.add(token)
            if token not in word_counts[label]:
                word_counts[label][token] = 0
            word_counts[label][token] += 1

    return word_counts, class_counts, vocabulary

# Predict using Naive Bayes
def predict_naive_bayes(text, word_counts, class_counts, vocabulary):
    tokens = tokenize(text)
    total_words = sum(class_counts.values())
    log_probabilities = {0: 0, 1: 0}

    for label in [0, 1]:
        log_probabilities[label] = class_counts[label] / total_words  # Prior probability
        for token in tokens:
            word_probability = (
                word_counts[label].get(token, 0) + 1
            ) / (class_counts[label] + len(vocabulary))  # Laplace smoothing
            log_probabilities[label] *= word_probability

    return max(log_probabilities, key=log_probabilities.get)

# Evaluate the model
def evaluate_model(X_test, y_test, word_counts, class_counts, vocabulary):
    correct = 0
    for text, label in zip(X_test, y_test):
        prediction = predict_naive_bayes(text, word_counts, class_counts, vocabulary)
        if prediction == label:
            correct += 1
    return correct / len(y_test)

# Main script
file_path = os.path.join(os.path.dirname(__file__), 'datasets', 'Hotel_Reviews.csv')
X, y = read_data(file_path)

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the model
word_counts, class_counts, vocabulary = train_naive_bayes(X_train, y_train)

# Evaluate on test set
accuracy = evaluate_model(X_test, y_test, word_counts, class_counts, vocabulary)

print(f'Accuracy: {accuracy:.2f}')

File: naive_bayes.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <unordered_set>
#include <algorithm>
#include <random>
#include <cmath>
#include <ctime>

// Function to read and preprocess data
void readData(const std::string &filePath, std::vector<std::string> &texts, std::vector<int> &labels) {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        std::cerr << "Error opening file: " << filePath << std::endl;
        return;
    }

    std::string line;
    bool isFirstLine = true;
    while (std::getline(file, line)) {
        if (isFirstLine) {
            isFirstLine = false;
            continue; // Skip header
        }

        std::stringstream ss(line);
        std::string negativeReview, positiveReview;
        std::getline(ss, negativeReview, ',');
        std::getline(ss, positiveReview, ',');

        if (negativeReview != "No Negative") {
            texts.push_back(negativeReview);
            labels.push_back(0);
        }

        if (positiveReview != "No Positive") {
            texts.push_back(positiveReview);
            labels.push_back(1);
        }
    }
    file.close();

    // Shuffle data
    std::srand(unsigned(std::time(nullptr)));
    std::vector<int> indices(texts.size());
    std::iota(indices.begin(), indices.end(), 0);

    std::mt19937 generator(std::random_device{}());
    std::shuffle(indices.begin(), indices.end(), generator);

    std::vector<std::string> shuffledTexts;
    std::vector<int> shuffledLabels;
    for (int idx : indices) {
        shuffledTexts.push_back(texts[idx]);
        shuffledLabels.push_back(labels[idx]);
    }
    texts = shuffledTexts;
    labels = shuffledLabels;
}

// Function to tokenize text
std::vector<std::string> tokenize(const std::string &text) {
    std::stringstream ss(text);
    std::string word;
    std::vector<std::string> tokens;
    while (ss >> word) {
        std::transform(word.begin(), word.end(), word.begin(), ::tolower);
        tokens.push_back(word);
    }
    return tokens;
}

// Function to train Naive Bayes classifier
void trainNaiveBayes(const std::vector<std::string> &X_train, const std::vector<int> &y_train,
                     std::map<int, std::map<std::string, int>> &wordCounts,
                     std::map<int, int> &classCounts,
                     std::unordered_set<std::string> &vocabulary) {
    for (size_t i = 0; i < X_train.size(); ++i) {
        auto tokens = tokenize(X_train[i]);
        classCounts[y_train[i]] += tokens.size();

        for (const auto &token : tokens) {
            vocabulary.insert(token);
            wordCounts[y_train[i]][token]++;
        }
    }
}

// Function to predict using Naive Bayes
int predictNaiveBayes(const std::string &text, const std::map<int, std::map<std::string, int>> &wordCounts,
                      const std::map<int, int> &classCounts, const std::unordered_set<std::string> &vocabulary) {
    auto tokens = tokenize(text);
    double totalWords = classCounts.at(0) + classCounts.at(1);
    std::map<int, double> logProbabilities;

    for (int label : {0, 1}) {
        logProbabilities[label] = std::log(classCounts.at(label) / totalWords);
        for (const auto &token : tokens) {
            double wordProbability = (wordCounts.at(label).count(token) ? wordCounts.at(label).at(token) + 1 : 1) /
                                     static_cast<double>(classCounts.at(label) + vocabulary.size());
            logProbabilities[label] += std::log(wordProbability);
        }
    }

    return logProbabilities[0] > logProbabilities[1] ? 0 : 1;
}

// Function to evaluate the model
double evaluateModel(const std::vector<std::string> &X_test, const std::vector<int> &y_test,
                     const std::map<int, std::map<std::string, int>> &wordCounts,
                     const std::map<int, int> &classCounts,
                     const std::unordered_set<std::string> &vocabulary) {
    int correct = 0;
    for (size_t i = 0; i < X_test.size(); ++i) {
        if (predictNaiveBayes(X_test[i], wordCounts, classCounts, vocabulary) == y_test[i]) {
            correct++;
        }
    }
    return static_cast<double>(correct) / X_test.size();
}

int main() {
    std::string filePath = "datasets/Hotel_Reviews.csv";
    std::vector<std::string> texts;
    std::vector<int> labels;

    // Read and preprocess data
    readData(filePath, texts, labels);

    // Split into train and test sets
    size_t trainSize = static_cast<size_t>(texts.size() * 0.8);
    std::vector<std::string> X_train(texts.begin(), texts.begin() + trainSize);
    std::vector<int> y_train(labels.begin(), labels.begin() + trainSize);
    std::vector<std::string> X_test(texts.begin() + trainSize, texts.end());
    std::vector<int> y_test(labels.begin() + trainSize, labels.end());

    // Train the model
    std::map<int, std::map<std::string, int>> wordCounts;
    std::map<int, int> classCounts;
    std::unordered_set<std::string> vocabulary;
    trainNaiveBayes(X_train, y_train, wordCounts, classCounts, vocabulary);

    // Evaluate on test set
    double accuracy = evaluateModel(X_test, y_test, wordCounts, classCounts, vocabulary);

    std::cout << "Accuracy: " << accuracy << std::endl;
    return 0;
}